proc_macros/src/lib.rs

```diff
@@ -165,3 +203,599 @@ pub fn rule_tests(input: TokenStream) -> TokenStream {
+                std::sync::Arc::new(#rule_instance_per_file_struct_name {
+                    rule_instance: self,
+                    #(#rule_instance_per_file_state_field_names: #rule_instance_per_file_state_field_initializers),*
+                })
+            }
+
+            fn rule(&self) -> Arc<dyn crate::rule::Rule> {
+                self.rule.clone()
+            }
+
+            fn listener_queries(&self) -> &[crate::rule::RuleListenerQuery] {
+                &self.listener_queries
+            }
+        }
+    }
+}
+
+fn get_rule_instance_per_file_struct_definition(
+    rule_instance_struct_name: &Ident,
+    rule_instance_per_file_struct_name: &Ident,
+    rule_instance_per_file_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let state_field_names = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.name);
+    let state_field_types = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.type_);
+    quote! {
+        struct #rule_instance_per_file_struct_name {
+            rule_instance: std::sync::Arc<#rule_instance_struct_name>,
+            #(#state_field_names: #state_field_types),*
+        }
+    }
+}
+
+struct SelfAccessRewriter<'a> {
+    rule: &'a Rule,
+}
+
+impl<'a> visit_mut::VisitMut for SelfAccessRewriter<'a> {
+    fn visit_expr_field_mut(&mut self, node: &mut ExprField) {
+        if let Some(self_field_name) = get_self_field_access_name(node) {
+            match self.rule.get_rule_state_scope_for_field(&self_field_name) {
+                Some(RuleStateScope::RuleStatic) => {
+                    let self_field_name = format_ident!("{}", self_field_name);
+                    *node = parse_quote!(self.rule_instance.rule.#self_field_name);
+                    return;
+                }
+                Some(RuleStateScope::PerRun) => {
+                    let self_field_name = format_ident!("{}", self_field_name);
+                    *node = parse_quote!(self.rule.#self_field_name);
+                    return;
+                }
+                _ => (),
+            }
+        }
+        visit_mut::visit_expr_field_mut(self, node);
+    }
+
+    fn visit_expr_macro_mut(&mut self, node: &mut ExprMacro) {
+        let parser = Punctuated::<Expr, Token![,]>::parse_terminated;
+        let rewritten_macro_args = match parser.parse2(node.mac.tokens.clone()) {
+            Ok(macro_args) => {
+                let rewritten_macro_args = macro_args
+                    .into_iter()
+                    .map(|mut macro_arg| {
+                        SelfAccessRewriter { rule: self.rule }.visit_expr_mut(&mut macro_arg);
+                        macro_arg
+                    })
+                    .collect::<Vec<_>>();
+                quote! {
+                    #(#rewritten_macro_args),*
+                }
+            }
+            _ => match syn::parse2::<ArrowSeparatedKeyValuePairs>(node.mac.tokens.clone()) {
```

Comment:
And if that didn't work, this handles parsing the macro contents as the `key => value` syntax (which was needed for `self.whatever` to work/get re-written inside a `violation!()` macro invocation)